### PR TITLE
[Enhancement] Improve registry infer_scope

### PR DESCRIPTION
**Pull request by zhouzaida**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

resolve https://github.com/open-mmlab/mmcv/issues/1116

- Before this PR

```python
>>> from mmengine import Registry
>>> CONVERTERS = Registry('converter')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/data/codebases/mmengine/mmengine/registry/registry.py", line 250, in __init__
    self._scope = self.infer_scope() if scope is None else scope
  File "/mnt/data/codebases/mmengine/mmengine/registry/registry.py", line 310, in infer_scope
    filename = inspect.getmodule(sys._getframe(2)).__name__  # type: ignore
AttributeError: 'NoneType' object has no attribute '__name__'
```

- After this PR

```python
>>> from mmengine import Registry
>>> CONVERTERS = Registry('converter')  # work as expected
```

## Modification

- Improve infer_scope at mmengine/registry/registry.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by HAOCHENYE — DISMISSED** [review]

File: mmengine/registry/registry.py [hunk 1]
```diff
@@ -307,9 +307,16 @@ def infer_scope() -> str:
...
-        return split_filename[0]
+        try:
+            filename = inspect.getmodule(  # type: ignore
+                sys._getframe(2)).__name__
+            split_filename = filename.split('.')
+            scope = split_filename[0]
+        except AttributeError:
+            # use "None" to handle some cases which can not infer the scope
+            # like initializing Registry in interactive mode
+            scope = 'None'
```
Comment (RangiLyu) [inline mmengine/registry/registry.py:318]: Should raise a warning here.

Comment (RangiLyu) [inline mmengine/registry/registry.py:318]: If the scope is None, will there be some problems when calling `build` or `get`? Especially when a none-scope node is added to a registry tree.

Comment (zhouzaida) [reply mmengine/registry/registry.py:159]: The scope is a `'None'` string rather than `None` object.

Comment (zhouzaida) [reply mmengine/registry/registry.py:161]: done

File: mmengine/registry/registry.py [hunk 2]
```diff
@@ -142,15 +142,31 @@ def infer_scope() -> str:
...
```
Comment (C1rN09) [inline mmengine/registry/registry.py:161]: Using `try ... except` may be ambiguous on what exactly happened. If this is intended to solve issue with interactive mode, maybe it's better to check `inspect.getmodule(...) is None` with an if-statement?

**Review by HAOCHENYE — APPROVED** [review]

File: mmengine/registry/registry.py (hunk 1 above)
Comment (zhouzaida) [reply mmengine/registry/registry.py]: I set it as 'mmengine'

File: mmengine/registry/registry.py (hunk 2 above)
Comment (zhouzaida) [reply mmengine/registry/registry.py]: Good suggestion. Done.

**Review by C1rN09 — APPROVED** [review]

**Review by RangiLyu — DISMISSED** [review]

File: mmengine/registry/registry.py [hunk 3]
```diff
@@ -142,15 +142,31 @@ def infer_scope() -> str:
...
+            filename = module.__name__
+            split_filename = filename.split('.')
+            scope = split_filename[0]
+        else:
+            # use "mmengine" to handle some cases which can not infer the scope
+            # like initializing Registry in interactive mode
+            scope = 'mmengine'
+            print_log(
+                'set scope as "mmengine" when scope can not be inferred. You '
+                'can disable this warning by passing a "scope" argument to '
```
Comment (zhouzaida) [inline mmengine/registry/registry.py:164]:
```suggestion
                'can silence this warning by passing a "scope" argument to '
```

**Pull request merged**